File: lib/rayz/shape.rb
require "matrix"

module Rayz
  class Shape
    attr_accessor :transform, :material, :parent
    attr_accessor :saved_ray # For test_shape debugging

    def initialize
      @transform = Matrix.identity(4)
      @material = Material.new
      @parent = nil
    end

    def intersect(ray)
      # Transform the ray by the inverse of the shape's transformation
      local_ray = ray.transform(@transform.inverse)
      @saved_ray = local_ray # For test_shape debugging
      local_intersect(local_ray)
    end

    def normal_at(world_point, hit = nil)
      # Transform world point to object space
      local_point = world_to_object(world_point)

      # Get the local normal
      local_normal = local_normal_at(local_point, hit)

      # Transform normal to world space
      normal_to_world(local_normal)
    end

    def world_to_object(point)
      # Convert point to object space by traversing parent hierarchy
      point = @parent.world_to_object(point) if @parent

      # Apply this object's inverse transformation
      object_point_matrix = @transform.inverse * point.to_matrix
      Point.new(
        x: object_point_matrix[0, 0],
        y: object_point_matrix[1, 0],
        z: object_point_matrix[2, 0]
      )
    end

    def normal_to_world(normal)
      # Apply this object's inverse transpose transformation
      world_normal_matrix = @transform.inverse.transpose * normal.to_matrix
      world_normal = Vector.new(
        x: world_normal_matrix[0, 0],
        y: world_normal_matrix[1, 0],
        z: world_normal_matrix[2, 0]
      )

      # Normalize the vector
      world_normal = world_normal.normalize

      # Continue transforming through parent hierarchy
      world_normal = @parent.normal_to_world(world_normal) if @parent

      world_normal
    end

    # Subclasses must implement these methods
    def local_intersect(local_ray)
      raise NotImplementedError, "Subclasses must implement local_intersect"
    end

    def local_normal_at(local_point, hit = nil)
      raise NotImplementedError, "Subclasses must implement local_normal_at"
    end

    def ==(other)
      return false unless other.class == self.class
      @transform == other.transform && @material == other.material
    end

    # Check if this shape includes another shape
    # Base implementation: a shape includes another if they are the same object
    def includes?(shape)
      self == shape
    end
  end
end

File: python/rayz/shape.py
# python/rayz/shape.py
from __future__ import annotations

from abc import ABC, abstractmethod

from rayz.material import Material
from rayz.matrix import Matrix
from rayz.tuple import Vector


class Shape(ABC):
    def __init__(self) -> None:
        self._transform = Matrix.identity(4)
        self._transform_inverse = Matrix.identity(4)
        self._transform_inverse_transpose = Matrix.identity(4)
        self.material = Material()
        self.parent = None
        # Callable[[float], Matrix] | None. If ever set to a lambda/closure,
        # it won't survive ProcessPoolExecutor pickling in render_parallel —
        # use a module-level function (see normal_perturbations.py's pattern)
        # if motion blur needs to work with the parallel renderer.
        self.motion_transform = None

    @property
    def transform(self) -> Matrix:
        return self._transform

    def set_transform(self, m: Matrix) -> None:
        self._transform = m
        self._transform_inverse = m.inverse()
        self._transform_inverse_transpose = m.inverse().transpose()
        self.invalidate_bounds_cache()

    def invalidate_bounds_cache(self) -> None:
        # No-op for shapes that don't cache their own bounds (only Group/CSG
        # do), but keeps propagating up so an ancestor Group/CSG's cache is
        # cleared whenever any descendant's transform changes.
        if self.parent is not None:
            self.parent.invalidate_bounds_cache()

    def intersect(self, ray) -> list:
        if self.motion_transform is not None:
            effective_inverse = (self.motion_transform(ray.time) * self._transform).inverse()
        else:
            effective_inverse = self._transform_inverse
        local_ray = ray.transform(effective_inverse)
        return self.local_intersect(local_ray)

    def world_to_object(self, point):
        p = point
        if self.parent is not None:
            p = self.parent.world_to_object(p)
        return self._transform_inverse * p

    def normal_to_world(self, normal) -> Vector:
        n = self._transform_inverse_transpose * normal
        n = Vector(n.x, n.y, n.z).normalize()
        if self.parent is not None:
            n = self.parent.normal_to_world(n)
        return n

    def normal_at(self, world_point, hit=None) -> Vector:
        local_point = self.world_to_object(world_point)
        local_normal = self.local_normal_at(local_point, hit)
        if self.material.normal_perturbation is not None:
            perturbation = self.material.normal_perturbation(local_point)
            local_normal = Vector(
                local_normal.x + perturbation.x,
                local_normal.y + perturbation.y,
                local_normal.z + perturbation.z,
            ).normalize()
        return self.normal_to_world(local_normal)

    def includes(self, shape) -> bool:
        return self is shape

    @abstractmethod
    def local_intersect(self, ray) -> list: ...

    @abstractmethod
    def local_normal_at(self, point, hit=None) -> Vector: ...

    @abstractmethod
    def bounds(self): ...


class TestShape(Shape):
    def __init__(self) -> None:
        super().__init__()
        self.saved_ray = None

    def local_intersect(self, ray) -> list:
        self.saved_ray = ray
        return []

    def local_normal_at(self, point, hit=None) -> Vector:
        return Vector(point.x, point.y, point.z)

    def bounds(self):
        from rayz.bounds import Bounds
        from rayz.tuple import Point

        return Bounds(Point(-1, -1, -1), Point(1, 1, 1))
